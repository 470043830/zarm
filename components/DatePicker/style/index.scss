@import "../../style/core/index";
@import "../../style/themes/default";
