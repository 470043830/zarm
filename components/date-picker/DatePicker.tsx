import React, { Component } from 'react';
import classnames from 'classnames';
import _ from 'lodash';
import BaseDatePickerProps from './PropsType';
import Popup from '../popup';
import DatePickerView from '../date-picker-view';
import removeFnFromProps from '../picker-view/utils/removeFnFromProps';
import { parseState } from '../date-picker-view/utils/parseState';

export interface DatePickerProps extends BaseDatePickerProps {
  prefixCls?: string;
  className?: string;
}

export default class DatePicker extends Component<DatePickerProps, any> {
  static defaultProps = {
    mode: 'date',
    value: '',
    defaultValue: '',
    minuteStep: 1,
    prefixCls: 'za-date-picker',
    valueMember: 'value',
    maskClosable: true,
    destroy: false,
    onCancel: () => {},
    onInit: () => {},
  };

  static getDerivedStateFromProps(props, state) {
    if (!_.isEqual(removeFnFromProps(props, ['onOk', 'onCancel', 'onChange']), removeFnFromProps(state.prevProps, ['onOk', 'onCancel', 'onChange']))) {
      return {
        prevProps: props,
        ...parseState(props),
      };
    }

    if (!_.isEqual(state.value, state.prevValue)) {
      return {
        prevValue: state.value,
        value: state.value,
      };
    }
    return null;
  }

  private initDate;

  private isScrolling = false;

  state = {
    value: '',
  };

  onCancel = () => {
    const { onCancel } = this.props;
    this.setState({
      value: this.initDate,
    });
    if (typeof onCancel === 'function') {
      onCancel();
    }
  };

  onOk = () => {
    if (this.isScrolling) {
      return false;
    }

    const { onOk } = this.props;
    const { value } = this.state;
    // const value = this.initDate;

    this.setState({
      value,
    });
    if (typeof onOk === 'function') {
      onOk(value);
    }
  };

  onTransition = (isScrolling) => {
    const { onTransition } = this.props;
    this.isScrolling = isScrolling;
    if (typeof onTransition === 'function') {
      onTransition(isScrolling);
    }
  };

  onInit = (selected) => {
    const { onInit } = this.props;
    // this.initDate = selected;
    this.setState({
      value: selected,
    });
    if (typeof onInit === 'function') {
      onInit(selected);
    }
  };

  onValueChange = (newValue) => {
    const { onChange } = this.props;
    this.setState({
      value: newValue,
    });

    if (typeof onChange === 'function') {
      onChange(newValue);
    }
  };

  close = (key) => {
    this.setState({
      [`${key}`]: false,
    });
  };

  render() {
    const { prefixCls, className, title, okText, cancelText, locale, getContainer, maskClosable, destroy, onOk, onCancel, onInit, visible, ...others } = this.props;
    const cls = classnames(prefixCls, className);
    const { value } = this.state;
    console.log('datePicker value:', value);
    const noop = () => {};

    return (
      <Popup
        visible={visible}
        onMaskClick={maskClosable ? this.onCancel : noop}
        getContainer={getContainer}
        destroy={destroy}
      >
        <div className={cls} onClick={(e) => { e.stopPropagation(); }}>
          <div className={`${prefixCls}__header`}>
            <div
              className={`${prefixCls}__cancel`}
              onClick={this.onCancel}
            >
              {cancelText || locale!.cancelText}
            </div>
            <div className={`${prefixCls}__title`}>{title || locale!.title}</div>
            <div className={`${prefixCls}__submit`} onClick={this.onOk}>{okText || locale!.okText}</div>
          </div>
          <DatePickerView
            {...others}
            className={className}
            value={value}
            onInit={this.onInit}
            onChange={this.onValueChange}
            onTransition={(isScrolling) => { this.onTransition(isScrolling); }}
          />
        </div>
      </Popup>
    );
  }
}
