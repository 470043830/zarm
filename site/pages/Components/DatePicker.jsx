import Markdown from '@site/components/Markdown';

export default class Page extends Markdown {
  document = () => {
    return {
      document: require('@/components/date-picker/demo.md'),
      className: 'datepicker-page',
    };
  }
}
