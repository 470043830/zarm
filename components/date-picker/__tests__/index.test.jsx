import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import DatePicker from '../index';

function fakeTimers() {
  const timer = jest.useFakeTimers();
  performance.timing = () => {};
}
const timer = fakeTimers();

describe('DatePicker', () => {
  it('DatePicker trigger maskClick', () => {
    const onMaskClickFn = jest.fn();
    const wrapper = mount(
      <DatePicker
        mode="date"
        value="2009-3-4"
        visible
        onMaskClick={onMaskClickFn}
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    // document.body.find('.za-mask').simulate('click');
    // expect(onMaskClickFn).toBeCalled();
  });

  it('should trigger onOk when press ok button', () => {
    const onOkFn = jest.fn();

    const wrapper = mount(
      <DatePicker
        mode="date"
        value="2009-3-4"
        visible
        onOk={onOkFn}
      />
    );
    wrapper.find('.za-date-picker__submit').simulate('click');
    expect(onOkFn).toBeCalled();
  });

  it('should trigger onCancel when press cancel button', () => {
    const onCancelFn = jest.fn();

    const wrapper = mount(
      <DatePicker
        mode="date"
        value="2009-3-4"
        visible
        onCancel={onCancelFn}
      />
    );

    wrapper.find('.za-date-picker__cancel').simulate('click');
    expect(onCancelFn).toBeCalled();
  });

  // it('should trigger onMaskClick when click mask', () => {
  //   const onMaskClick = jest.fn();

  //   const wrapper = mount(
  //     <DatePicker
  //       mode="date"
  //       value="2009-3-4"
  //       visible
  //       onMaskClick={onMaskClick}
  //     />
  //   );

  //   wrapper.find('.za-mask').simulate('click');
  //   expect(onMaskClick).toBeCalled();
  // });
});
