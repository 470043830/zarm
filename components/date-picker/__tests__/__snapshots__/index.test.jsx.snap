// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker DatePicker trigger maskClick 1`] = `
<DatePicker
  cancelText="取消"
  defaultValue=""
  disabled={false}
  minuteStep={1}
  mode="date"
  okText="确定"
  onCancel={[Function]}
  onInit={[Function]}
  onMaskClick={[MockFunction]}
  placeholder="请选择"
  prefixCls="za-picker"
  title="请选择"
  value="2009-3-4"
  valueMember="value"
  visible={true}
>
  <Popup
    onMaskClick={[Function]}
    visible={true}
  />
</DatePicker>
`;
